raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java

```diff
@@ -686,21 +730,41 @@ private boolean handleNonMatchingResponseLeaderAndEpoch(int epoch, OptionalInt l
         return false;
     }
 
-    private void handleResponse(RaftResponse.Inbound response) throws IOException {
+    private void handleResponse(RaftResponse.Inbound response, long currentTimeMs) throws IOException {
         // The response epoch matches the local epoch, so we can handle the response
-        ApiMessage responseData = response.data();
-        if (responseData instanceof FetchQuorumRecordsResponseData) {
-            handleFetchQuorumRecordsResponse((FetchQuorumRecordsResponseData) responseData);
-        } else if (responseData instanceof VoteResponseData) {
-            handleVoteResponse(response.sourceId(), (VoteResponseData) responseData);
-        } else if (responseData instanceof BeginQuorumEpochResponseData) {
-            handleBeginQuorumEpochResponse(response.sourceId(), (BeginQuorumEpochResponseData) responseData);
-        } else if (responseData instanceof EndQuorumEpochResponseData) {
-            // Nothing to do for now
-        } else if (responseData instanceof FindQuorumResponseData) {
-            handleFindQuorumResponse((FindQuorumResponseData) responseData);
+        ApiKeys apiKey = ApiKeys.forId(response.data.apiKey());
+        final Optional<Errors> errorOpt;
+
+        switch (apiKey) {
+            case FETCH_QUORUM_RECORDS:
+                errorOpt = handleFetchQuorumRecordsResponse(response);
+                break;
+
+            case VOTE:
+                errorOpt = handleVoteResponse(response);
+                break;
+
+            case BEGIN_QUORUM_EPOCH:
+                errorOpt = handleBeginQuorumEpochResponse(response);
+                break;
+
+            case END_QUORUM_EPOCH:
+                errorOpt = handleEndQuorumEpochResponse(response);
+                break;
+
+            case FIND_QUORUM:
+                errorOpt = handleFindQuorumResponse(response);
+                break;
+
+            default:
+                throw new IllegalArgumentException("Received unexpected response type: " + apiKey);
+        }
+
+        ConnectionState connection = connections.getOrCreate(response.sourceId());
+        if (errorOpt.isPresent()) {
```

Comment:
If we check `errorOpt.isPresent() && errorOpt.get() != NONE` here, then we would not need the special handling in fetch quorum records of `handleNonMatchingResponseLeaderAndEpoch` right?

Reply:
I was using the `Optional` type here to represent the presence of absence of an error, so the need to check for `NONE` would defeat that. Let me think a little about this. Perhaps we should just return `Errors` directly.